crypto/store/store_lib.c

```diff
@@ -0,0 +1,396 @@
+        ret = ctx->loader->ctrl(ctx->loader_ctx, cmd, args);
+    va_end(args);
+
+    return ret;
+}
+
+STORE_INFO *STORE_load(STORE_CTX *ctx)
+{
+    STORE_INFO *v = NULL;
+
+ again:
+    v = ctx->loader->load(ctx->loader_ctx, ctx->ui_method, ctx->ui_data);
+
+    if (ctx->post_process != NULL && v != NULL) {
+        v = ctx->post_process(v, ctx->post_process_data);
+
+        /*
+         * By returning NULL, the callback decides that this object should
+         * be ignored.
+         */
+        if (v == NULL)
+            goto again;
+    }
+
+    return v;
+}
+
+int STORE_eof(STORE_CTX *ctx)
+{
+    return ctx->loader->eof(ctx->loader_ctx);
+}
+
+int STORE_close(STORE_CTX *ctx)
+{
+    int loader_ret = ctx->loader->close(ctx->loader_ctx);
+
+    OPENSSL_free(ctx);
+    return loader_ret;
+}
+
+/*
+ * Functions to generate STORE_INFOs, one function for each type we
+ * support having in them.  Along with each of them, one macro that
+ * can be used to determine what types are supported.
+ *
+ * In all cases, ownership of the object is transfered to the STORE_INFO
+ * and will therefore be freed when the STORE_INFO is freed.
+ */
+static STORE_INFO *store_info_new(int type, void *data)
+{
+    STORE_INFO *info = OPENSSL_zalloc(sizeof(*info));
+
+    if (info == NULL)
+        return NULL;
+
+    info->type = type;
+    info->_.data = data;
+    return info;
+}
+
+STORE_INFO *STORE_INFO_new_NAME(char *name)
+{
+    STORE_INFO *info = store_info_new(STORE_INFO_NAME, NULL);
+
+    if (info == NULL) {
+        STOREerr(STORE_F_STORE_INFO_NEW_NAME, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    info->_.name.name = name;
+    info->_.name.desc = NULL;
+
+    return info;
+}
+
+int STORE_INFO_set0_NAME_description(STORE_INFO *info, char *desc)
+{
+    if (info->type != STORE_INFO_NAME) {
+        STOREerr(STORE_F_STORE_INFO_SET0_NAME_DESCRIPTION,
+                 ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+
+    info->_.name.desc = desc;
+
+    return 1;
+}
+STORE_INFO *STORE_INFO_new_PARAMS(EVP_PKEY *params)
+{
+    STORE_INFO *info = store_info_new(STORE_INFO_PARAMS, params);
+
+    if (info == NULL)
+        STOREerr(STORE_F_STORE_INFO_NEW_PARAMS, ERR_R_MALLOC_FAILURE);
+    return info;
+}
+
+STORE_INFO *STORE_INFO_new_PKEY(EVP_PKEY *pkey)
+{
+    STORE_INFO *info = store_info_new(STORE_INFO_PKEY, pkey);
+
+    if (info == NULL)
+        STOREerr(STORE_F_STORE_INFO_NEW_PKEY, ERR_R_MALLOC_FAILURE);
+    return info;
+}
+
+STORE_INFO *STORE_INFO_new_CERT(X509 *x509)
+{
+    STORE_INFO *info = store_info_new(STORE_INFO_CERT, x509);
+
+    if (info == NULL)
+        STOREerr(STORE_F_STORE_INFO_NEW_CERT, ERR_R_MALLOC_FAILURE);
+    return info;
+}
+
+STORE_INFO *STORE_INFO_new_CRL(X509_CRL *crl)
+{
+    STORE_INFO *info = store_info_new(STORE_INFO_CRL, crl);
+
+    if (info == NULL)
+        STOREerr(STORE_F_STORE_INFO_NEW_CRL, ERR_R_MALLOC_FAILURE);
+    return info;
+}
+
+STORE_INFO *STORE_INFO_new_ENDOFDATA(void)
```

Comment:
Don't like this naming "ENDOFDATA" ... that isn't following the pattern of the other items - you are not collecting "ENDOFDATA" objects.